docs/src/DeveloperDocumentation/styleguide.md

```diff
@@ -241,3 +217,189 @@ existing block.
+
+# supercompact
+General linear group
+```
+
+The print modes are specified as follows
+#### Detailed printing
+- the output must make sense as a standalone without context to non-specialists
+- the number of output lines should fit in the terminal
+- if the object is simple enough use only one line
+- use indentation and (usually) `:oneline` to print substructures
+#### One line printing
+- the output must print in one line
+- should make sense as a standalone without context
+- variable names/generators/relations should not be printed only their number.
+- Only the first word is capitalized e.g. `Polynomial ring`
+- one should use `:supercompact` for nested printing in compact
+- nested calls to `:oneline` (if you think them really necessary) should be at the end,
+  so that one can read sequentially. Calls to `:supercompact` can be anywhere.
+#### Super compact printing
+- no nested printing
+- a single term or a symbol/letter mimicking mathematical notation
+- should usually only depend on the type and not of the type parameters or of
+  the concrete instance - exceptions of this rule are possible e.g. for `GF(2)`
+- no nested printing. In particular variable names and `base_ring` must not be displayed.
+  This ensures that `:oneline` and `:supercompact` stay compact even for complicated things.
+  If you want nested printing use `:oneline` or `details`.
+- a user readable version of the main (mathematical) type.
+
+### Implementing show functions
+
+Here is the translation between `:detail`, `:oneline` and `:supercompact`.
+
+```
+print(io, "text/plain", x)                 # detailed printing
+print(io, x)                               # oneline printing
+print(IOContext(:supercompact => true), x) # supercompact printing
+```
+
+For reference, string interpolation `"$(x)"` will also use `print(io, x)`.
+
+#### Mockup
+
+```julia
+struct NewRing
+  base_ring
+end
+```
+
+```julia
+function Base.show(io::IO, ::MIME"text/plain", R::NewRing)
+    println(io, "I am a new ring")
+    println(io, "I print with newlines")
+    print(io, R.base_ring)
+end
+```
+
+```julia
+function Base.show(io::IO, R::NewRing)
+    if get(io, :supercompact, false)
+        print(io, "supercompact printing of newring ")
+        print(io, R.base_ring) # this prints the :oneline version of newring
+    else
+        print(io, "oneline printing of newring with ")
+        print(IOContext(io, :supercompact => true), "supercompact ", R.base_ring)
+    end
+end
+```
+
+
+#### Alternative
+
+Here is an alternative version, which needs to be used in case the detailed
+printing does not contain newlines.
+
+```julia
+function Base.show(io::IO, R::NewRing)
+    if get(io, :supercompact, false)
+      print(io, "supercompact printing of newring")
+    else
+      println(io, "I am a new ring and always print in one line" )
+      print(IOContext(io, :supercompact => true), base_ring(R))
+    end
+end
+```
+
+#### Not working as expected
```

Comment:
I don't really understand from this, what is not working as expected here and why. Also, is the "Alternative" section above a solution to this problem or is it unrelated?

Reply:
You are right, all the information is there, but it took me quite a while to put it together in the right order.
Maybe there could be a short explanation "Then the following will *not* be used for array/tuple printing, BECAUSE ..." to make it clearer, why the "Not working as expected"-version fails.
Also, in the "Alternative" section, it wasn't immediately clear to me that the `else` case in fact covers both detailed and oneline printing (so that the function replaces both `show` functions from before).

Reply:
I added a few explanations.

Reply:
Thank you!